core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -305,24 +394,12 @@ class half_t {
   }
 
   KOKKOS_FUNCTION
-  volatile half_t& operator+=(half_t rhs) volatile {
-#ifdef __CUDA_ARCH__
-    // Cuda 10 supports __half volatile stores but not volatile arithmetic
-    // operands. Cast away volatile-ness of val for arithmetic but not for store
-    // location.
-    val = const_cast<impl_type&>(val) + rhs.val;
-#else
-    // Use non-volatile val_ref to suppress:
-    // "warning: implicit dereference will not access object of type ‘volatile
-    // __half’ in statement"
-    auto val_ref = const_cast<impl_type&>(val);
-    val_ref      = __float2half(__half2float(const_cast<impl_type&>(val)) +
-                           __half2float(rhs.val));
-#endif
+  volatile half_t operator+=(const volatile half_t rhs) volatile {
+    atomic_add(rhs);
```

Comment:
Why does this need to be `atomic`?

Reply:
Same as above. See https://github.com/kokkos/kokkos/pull/4101#discussion_r681229041.